Allow empty buffer slices to be instantiated

beholdnec opened the pull request:
**Connections**
Addresses #6779. Some developers would like for wgpu not to panic when 0-length buffer slices exist.

**Description**
wgpu currently panics if an empty buffer slice (length 0) is created. The current code has contradictory documentation about whether empty buffer slices are allowed to exist.

This PR allows empty buffer slices to exist.

Empty buffer slices can be:
- Instantiated
- Mapped (the result is an empty slice)

Empty buffer slices cannot be:
- Used in buffer bindings
- Used in `set_index_buffer` or `set_vertex_buffer` (presently -- this is allowed by WebGPU, but not supported in `wgpu`. See #3170.)

**Testing**
Tests in buffer.rs have been modified to accept zero-length buffer slices. All tests pass.

**Checklist**

- [X] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [X] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [ ] Run `cargo xtask test` to run tests.
- [X] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

beholdnec committed:
Allow empty buffer slices to be instantiated

beholdnec committed:
Allow empty buffer slices to be instantiated

beholdnec commented:
Open questions:
- should we silently accept empty slices instead of panicking? If so, which methods should accept empty slices?
- an empty buffer slice could be a logic error, and this PR can potentially make such errors less detectable. Is this still desirable for developers?

andyleiserson commented:
Related: #3170 

The WebGPU spec allows zero-size bindings in places where we currently don't. I think that answers the question of whether we should accept them, but we need to figure out how to handle them internally when the backend does not support them.

inner-daemons commented:
@beholdnec Any updates on this?

beholdnec commented:
> @beholdnec Any updates on this?

Not really, I was kind of hoping there would be more discussion. See "Open questions" above.

beholdnec committed:
Allow mapping 0-sized buffer ranges

beholdnec committed:
Allow mapping 0-sized buffer ranges

beholdnec committed:
Update changelog

beholdnec committed:
Update changelog

beholdnec committed:
Fix GL 0-sized map workaround

beholdnec committed:
Fix GL 0-sized map workaround

beholdnec commented:
Added some tweaks to allow mapping 0-sized ranges, which is supported by WebGPU.

beholdnec committed:
Fix if GLES buffer mapped for length outside of i32 range

beholdnec committed:
Fix if GLES buffer mapped for length outside of i32 range

beholdnec committed:
Improve error handling on GLES mapped buffer range

beholdnec committed:
Improve error handling on GLES mapped buffer range

beholdnec committed:
Buffer mappings no longer have to be non-zero

beholdnec committed:
Buffer mappings no longer have to be non-zero

andyleiserson commented on wgpu/src/api/buffer.rs:
```diff
@@ -720,9 +724,9 @@ struct Subrange {
 
 impl Subrange {
-    fn new(offset: BufferAddress, size: BufferSize, kind: RangeMappingKind) -> Self {
+    fn new(offset: BufferAddress, size: BufferAddress, kind: RangeMappingKind) -> Self {
```
While we're changing this, I'd propose renaming it to `new_unchecked` and documenting the requirement that `offset + size` does not overflow.

beholdnec replied:
Do you mean keeping `new` with `BufferSize`, and adding an additional method `new_unchecked` that takes `BufferAddress`? Should `new_unchecked` be marked unsafe?

andyleiserson replied:
Sorry for the confusion. I said "while we're changing this" because the suggestion isn't related to `BufferAddress` vs. `BufferSize`, or even to your change at all. It's just a small cleanup I noticed. The issue is that the addition `offset + size` does not check for overflow. Since this is a non-`pub` internal helper, I was suggesting renaming the method to `new_unchecked` to serve as a warning when calling it that the caller is asserting the combination of `offset` and `size` is valid. There isn't directly a risk of UB here, so it doesn't need to be `unsafe`.

Don't worry about it if you'd rather not do this in your PR, it's not a concern about the PR itself, so I'm fine merging the PR without addressing this.

andyleiserson commented on wgpu/src/api/render_bundle_encoder.rs:
```diff
@@ -93,7 +93,7 @@ impl<'a> RenderBundleEncoder<'a> {
             index_format,
             buffer_slice.offset,
-            Some(buffer_slice.size),
+            Some(buffer_slice.size_expect_nonzero()),
```
I guess a panic here is not a new panic, it's just the panic that existed before (constructing an empty `BufferSlice`) moved elsewhere?

Maybe add `TODO(https://github.com/gfx-rs/wgpu/issues/3170): Empty slices should be supported here`.

andyleiserson commented on wgpu/src/api/buffer.rs:
```diff
@@ -674,7 +678,7 @@ impl<'a> From<BufferSlice<'a>> for crate::BufferBinding<'a> {
             buffer: value.buffer,
             offset: value.offset,
-            size: Some(value.size),
+            size: Some(value.size_expect_nonzero()),
```
Since `BufferBinding` cannot be empty, this error case (unlike the one for vertex/index buffers) is going to stick around, and I think it would be better to avoid the panic, even if it requires a breaking change (the most obvious candidate is changing this to a `TryFrom`).

We recently changed `wgpu_core::BufferBinding` to support both `None` (size spans remainder of buffer) and `Some(0)` (explicit zero size), so that `wgpu-core` can validate the JS API. Possibly we should just change `wgpu::BufferBinding` to work the same way, and the application will get an error in `create_bind_group` if the range is empty.

Alternatively, we could introduce something like `NonEmptyBufferSlice`, `NonEmpty<BufferSlice>`, or `BufferSlice<S: Into<BufferAddress> = BufferSize>`.

Or I guess the way to do it without a breaking change is to keep `Buffer::slice` and `BufferSlice` the same, and add something new for when it is permitted to be empty.

beholdnec replied:
I changed From to TryFrom for BufferSlice -> BufferBinding.

andyleiserson reviewed (changes requested):
I'd like to iterate on the API a bit more, hopefully we can come up with something that eliminates some of the panic cases entirely.

It doesn't make sense to do until the design is finalized, but there are also some documentation updates needed (e.g. `hal::Device::map_buffer` and `hal::BufferMapping`), and there should be some end-to-end tests for zero-size mappings.

beholdnec committed:
Fix compiler errors from rebase

beholdnec committed:
Fix compiler errors from rebase

beholdnec committed:
Change From to TryFrom for converting BufferSlice to BufferBinding

beholdnec committed:
Change From to TryFrom for converting BufferSlice to BufferBinding

beholdnec committed:
Fix documentation

beholdnec committed:
Fix documentation

andyleiserson commented on CHANGELOG.md:
```diff
@@ -52,6 +52,7 @@ Bottom level categories:
 - Make `wgpu_types::texture::format::TextureChannel` accessible as `wgpu::TextureChannel`. By @TornaxO7 in [#9394](https://github.com/gfx-rs/wgpu/pull/9349).
 - Add support for `per_vertex` in Metal and DX12, as well as some validation for `per_vertex`, and a new enable extension, `wgpu_per_vertex`. By @inner-daemons in [#9219](https://github.com/gfx-rs/wgpu/pull/9219).
 - Add `ComputePass` version of `CommandEncoder::transition_resources` that allows intra-pass transitions. By @wingertge in [#9371](https://github.com/gfx-rs/wgpu/pull/9371).
+- Buffer slices with a length of 0 can now be instantiated. By @beholdnec in [#8505] (https://github.com/gfx-rs/wgpu/pull/8505).
```
Let's move this to the "major changes" section of the changelog. I'd suggest adding the following:

> #3170 tracks making it possible to pass a zero-size `BufferSlice` to `set_vertex_buffer` and `set_index_buffer` in the future.
> 
> Zero-size buffer bindings are still not permitted. `BufferBinding` now implements `TryFrom<BufferSlice>` instead of `From<BufferSlice>`. The `TryFrom` conversion will fail if the slice is zero-size.
> 
> ```diff
> -let slice = buffer.slice(0..0); // panic!
> -let mapping = BufferBinding::from(slice); // infallible
> +let slice = buffer.slice(0..0); // okay
> +let mapping = BufferBinding::try_from(slice).unwrap(); // panic
> ```

andyleiserson commented on wgpu/src/api/render_pass.rs:
```diff
@@ -105,7 +105,7 @@ impl RenderPass<'_> {
             index_format,
             buffer_slice.offset,
-            Some(buffer_slice.size),
+            Some(buffer_slice.size_expect_nonzero()),
```
nit: The TODO(https://github.com/gfx-rs/wgpu/issues/3170) comment also applies to the two in this file.

andyleiserson reviewed (changes requested):
I noticed that the buffer mapping API (`wgpu/src/api/buffer.rs`) interprets `MapContext.mapped_range` of `0..0` as "not mapped", this should be changed to a different sentinel value (or maybe changed to `Option`, I'm not sure it's really worth trying to save a few bytes here).

I'd still like to have the documentation updates I mentioned in my last review and a functional test of zero-size buffer mapping. (This might be covered by the CTS, if so, then we should enable the relevant test in `cts_runner/test.lst`.)

beholdnec committed:
Address PR feedback

beholdnec committed:
Address PR feedback

beholdnec committed:
Don't use an empty slice as a sentinel value

beholdnec committed:
Don't use an empty slice as a sentinel value

beholdnec committed:
Fixup empty buffer tests; fix empty slice mapping in GLES; changelog

beholdnec committed:
Fixup empty buffer tests; fix empty slice mapping in GLES; changelog

beholdnec commented:
Added some commits to address PR feedback.

I fixed up the "empty_buffer" test and enabled it for buffers with READ usage (it fails READ | WRITE usage for reasons unrelated to this PR). Tests pass.

beholdnec committed:
Refine empty buffer test

beholdnec committed:
Refine empty buffer test

beholdnec committed:
Fix potential errors in GLES

beholdnec committed:
Fix potential errors in GLES

Wumpf commented on wgpu/src/api/buffer.rs:
```diff
@@ -662,28 +667,38 @@ impl<'a> BufferSlice<'a> {
 }
 
-impl<'a> From<BufferSlice<'a>> for crate::BufferBinding<'a> {
+impl<'a> TryFrom<BufferSlice<'a>> for crate::BufferBinding<'a> {
```
`BufferSlice`'s documentation notes down the previous `From` impl, writing `You can also convert it to a [`BufferBinding`] with .into().` which is no longer the case. Needs to be adjusted and point out the zero size limitations there.

Wumpf commented on wgpu/src/api/buffer.rs:
we now no longer panic there. That's a validation error now :)

Wumpf commented on CHANGELOG.md:
````diff
@@ -62,6 +62,33 @@ This allows gaps in `VertexState`'s `buffers` and adds support for unbinding ver
++let mapping = BufferBinding::try_from(slice).unwrap(); // panic
+```
+
+By @beholdnec in [#8505] (https://github.com/gfx-rs/wgpu/pull/8505).
````
```suggestion
By @beholdnec in [#8505](https://github.com/gfx-rs/wgpu/pull/8505).
```

Wumpf reviewed (changes requested):
Looks great, I think this is a looot better than before, reducing the amount of surprise-unwraps inside of wgpu!

Found a few places where docs need to be adapted a bit further. Otherwise good to go!

All the wgpu methods that call internally `size_expect_nonzero` now need to also point out this way of panicing - the panic happened earlier in the process before, so it has shifted to there.
(those can come with the same TODO note as in other places where we should actually support it)

beholdnec committed:
Address PR feedback

beholdnec commented:
@Wumpf Pushed a commit to address PR feedback.

Wumpf reviewed (approved):
🚢
